lib/icinga/scheduleddowntime.cpp

```diff
@@ -292,6 +310,24 @@ void ScheduledDowntime::CreateNextDowntime()
 	}
 }
 
+void ScheduledDowntime::RemoveObsoleteDowntimes()
+{
+	auto name (GetName());
+	auto downtimeOptionsHash (HashDowntimeOptions());
+
+	// Just to be sure start and removal don't happen at the same time
+	auto threshold (Utility::GetTime() + 5 * 60);
+
+	for (const Downtime::Ptr& downtime : GetCheckable()->GetDowntimes()) {
+		if (downtime->GetScheduledBy() == name && downtime->GetStartTime() > threshold) {
+			auto configOwnerHash (downtime->GetConfigOwnerHash());
+
+			if (!configOwnerHash.IsEmpty() && configOwnerHash != downtimeOptionsHash)
```

Comment:
What was the intention of not re-creating downtimes that don't have the config owner hash set?

Reply:
Imagine you have ScheduledDowntimes and DowntimeStart notifications everywhere. And then you upgrade...

Reply:
But since you only delete downtimes that start in more than 5 minutes, no start notifications have been sent yet. So I think the following should work fine:
1. Remove the `!configOwnerHash.IsEmpty()` here -> delete every downtime with no or a wrong hash starting more that 5 minutes in the future
2. Leave the check in `CreateNextDowntime()` as it is right now -> don't reschedule if there's still a downtime without a hash set
3. In `TimerProc()`, call `RemoveObsoleteDowntimes()` before `CreateNextDowntime()` to minimize the time between deleting the old and adding the new downtime

Reply:
I'm afraid that doesn’t matter for the end notification.

```
[2021-07-02 17:05:51 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!48db09bb-1ea6-40d1-8bac-d56c57ddfd15' of type 'Downtime'.
[2021-07-02 17:05:51 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!48db09bb-1ea6-40d1-8bac-d56c57ddfd15' between '2021-07-02 17:07:00' and '2021-07-02 19:30:00', author: 'me', fixed
[2021-07-02 17:05:51 +0200] information/Checkable: Checkable 'Alexanders-MacBook-Pro.local' has 1 notification(s). Checking filters for type 'DowntimeEnd', sends will be logged.
[2021-07-02 17:05:51 +0200] information/Notification: Sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for user 'icingaadmin'
[2021-07-02 17:05:51 +0200] information/ConfigObjectUtility: Deleted object 'Alexanders-MacBook-Pro.local!11f091fb-d434-48d7-bb90-4aa76614dfcd' of type 'Downtime'.
[2021-07-02 17:05:51 +0200] information/Notification: Completed sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for checkable 'Alexanders-MacBook-Pro.local' and user 'icingaadmin' using command 'mail-host-notification'.
[2021-07-02 17:05:51 +0200] information/Downtime: Removed downtime 'Alexanders-MacBook-Pro.local!11f091fb-d434-48d7-bb90-4aa76614dfcd' from checkable 'Alexanders-MacBook-Pro.local' (Reason: expired at 2021-07-02 19:30:00 +0200).
```

Reply:
Wait what, end notifications are also sent for downtimes that weren't even started?

Reply:
Should be fixed by #8874